File: 0286-walls-and-gates/0286-walls-and-gates.py
from collections import deque

class Solution(object):
    def wallsAndGates(self, rooms):
        """
        :type rooms: List[List[int]]
        :rtype: None Do not return anything, modify rooms in-place instead.
        """
        EMPTY = float('inf')
        GATE = 0
        DIRECTIONS = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        if not rooms:
            return
        
        m = len(rooms)
        n = len(rooms[0])
        queue = deque()

        # Add all gates to the queue
        for row in range(m):
            for col in range(n):
                if rooms[row][col] == GATE:
                    queue.append((row, col))
        
        # Perform BFS
        while queue:
            row, col = queue.popleft()
            current_distance = rooms[row][col]
            for dr, dc in DIRECTIONS:
                r, c = row + dr, col + dc
                if 0 <= r < m and 0 <= c < n and rooms[r][c] == 2147483647:
                    rooms[r][c] = current_distance + 1
                    queue.append((r, c))

File: 0286-walls-and-gates/0286-walls-and-gates.java
class Solution {
    //a. Some variable initializations here
    public Queue<int[]> queue = new LinkedList<>();
    public int ROWS = 0;
    public int COLS = 0;
    
    public int[][] directions = new int[][]{{-1,0},{0,1},{1,0},{0,-1}};
    public void wallsAndGates(int[][] rooms) {
        //1. Initialize the variables 
        this.ROWS = rooms.length;
        this.COLS = rooms[0].length;
        
        //2. For loop 
        for(int r = 0;r<this.ROWS;r++){
            for(int c = 0;c<this.COLS;c++){
                //a. find if its a gate then put in queue
                if(rooms[r][c] == 0){
                    this.queue.offer(new int[]{r,c});
                }
            }
        }
        
        //3. call the bfs function
        this.bfs(rooms);
        
        //4. no return
    }
    
    public void bfs(int[][] rooms){
        //1. Initialize the queue with some stuff (not needed)
        //2. While queue 
        while(!this.queue.isEmpty()){
            //a. Pop the first element 
            
            int[] current= this.queue.poll();
            int dist = rooms[current[0]][current[1]];
            //b. loop through all directions 
            for(int[] direction: this.directions){
                //c. update the values 
                int newR = current[0] + direction[0];
                int newC = current[1]+ direction[1];
                
                //d. check the conditoin 
                if((newR>=0) &&(newR<this.ROWS) &&(newC>=0)&&(newC<this.COLS) && (rooms[newR][newC] == 2147483647)){
                    rooms[newR][newC] = dist+1;
                    this.queue.offer(new int[]{newR, newC});
                    
                    
                }
            }
        }
    }
}